talk page - CoreData housekeeping

tonisevener opened the pull request:
Delete talk pages > 50

tonisevener commented on Wikipedia/Code/TalkPageLocalHandler.swift:
```diff
@@ -124,7 +135,7 @@ extension NSManagedObjectContext {
         let request: NSFetchRequest<TalkPageTopicContent> = TalkPageTopicContent.fetchRequest()
         request.predicate = NSPredicate(format: "topics.@count == 0")
-        let results = try request.execute()
+        let results = try fetch(request)
```
before this change I was getting a "Cannot fetch without an NSManagedObjectContext in scope}" console error and TalkPageTopicContent objects were not clearing out in housekeeping.

tonisevener commented on WMF Framework/Wikipedia.xcdatamodeld/Wikipedia 2.xcdatamodel/contents:
```diff
@@ -70,6 +70,7 @@
         </fetchIndex>
     </entity>
     <entity name="TalkPage" representedClassName="TalkPage" syncable="YES" codeGenerationType="class">
+        <attribute name="dateAccessed" attributeType="Date" defaultDateTimeInterval="582233760" usesScalarValueType="NO" syncable="YES"/>
```
This defaults to the date/time I created the attribute (06/14/2019), let me know if there's a better one.

nambatee commented on Wikipedia/Code/WMFDatabaseHouseKeeper.swift:
```diff
@@ -6,9 +6,31 @@ import Foundation
+        
+        let oldTalkPages = try moc.fetch(request)
+        for talkPage in oldTalkPages {
+            moc.delete(talkPage)
```
might be a good use case for a batch delete https://developer.apple.com/library/archive/featuredarticles/CoreData_Batch_Guide/BatchDeletes/BatchDeletes.html

joewalsh commented on Wikipedia/Code/TalkPageLocalHandler.swift:
```diff
@@ -20,14 +20,23 @@ extension NSManagedObjectContext {
         
-        return try fetch(fetchRequest).first
+        let talkPage = try fetch(fetchRequest).first
+        talkPage?.dateAccessed = Date()
```
I'm a bit concerned that the method for fetching a talk page also mutates it and saves the context - could I move this into a separate function that's explicitly called when the user views the talk page or is there some other consideration that I'm missing?

joewalsh replied:
I think I misread "accessed" to mean "viewed by the user" - is the intent to have this property updated anytime the object is accessed even by non-user task?

tonisevener replied:
@joewalsh true, agreed, non-user task shouldn't update this in my opinion. I really meant viewed by the user & wanted to keep track of what talk pages they seem to be interested in lately and update that date. Then later I sort by that date so we are deleting talk pages that they don't seem to be visiting anymore. To keep it simpler we could probably only timestamp this after it's fetched in the view controller so we know it was instigated by the user.

joewalsh replied:
@tonisevener makes sense. I pushed a change, but feel free to alter or revert if you had other ideas

tonisevener commented:
Updated PR feedback -

1. I removed dateAccessed timestamp from TalkPageLocalHandler, I think only setting it in TalkPageContainerVC should be enough.
2. Added batch delete in `deleteStaleTalkPages` & `removeUnlinkedTalkPageTopicContent` and am merging the changes in case the user has that talk page in memory. I didn't bother with reacting on the UI-side to a talk page deleting. I don't really see how a talk page currently on screen could become the 50th talk page sorted by access date and delete out from under them in housekeeping, so it seems unnecessary. But, I'm happy to handle if you all think otherwise.
